crates/core/src/factories/core_fetch.rs

```diff
@@ -348,40 +352,53 @@ impl Drop for CoreFetch {
 }
 
 #[derive(Debug)]
-struct CoolDownList {
-    state: HashMap<AgentId, Instant>,
-    cool_down_interval: u64,
+struct BackOffList {
+    state: HashMap<AgentId, (Instant, u32)>,
+    back_off_interval: u64,
+    max_back_off_exponent: u32,
 }
 
-impl CoolDownList {
-    pub fn new(cool_down_interval: u64) -> Self {
+impl BackOffList {
+    pub fn new(back_off_interval: u64, max_back_off_exponent: u32) -> Self {
         Self {
             state: HashMap::new(),
-            cool_down_interval,
+            back_off_interval,
+            max_back_off_exponent,
         }
     }
 
-    pub fn add_agent(&mut self, agent_id: AgentId) {
-        self.state.insert(agent_id, Instant::now());
+    pub fn back_off_agent(&mut self, agent_id: &AgentId) {
+        match self.state.entry(agent_id.clone()) {
+            Entry::Occupied(mut o) => {
+                o.get_mut().0 = Instant::now();
+                o.get_mut().1 = self.max_back_off_exponent.min(o.get().1 + 1);
+            }
+            Entry::Vacant(v) => {
+                v.insert((Instant::now(), 0));
+            }
+        }
     }
 
-    pub fn is_agent_cooling_down(&mut self, agent_id: &AgentId) -> bool {
+    pub fn is_agent_backing_off(&mut self, agent_id: &AgentId) -> bool {
         match self.state.get(agent_id) {
-            Some(instant) => {
-                if instant.elapsed().as_millis()
-                    > self.cool_down_interval as u128
-                {
-                    // Cool down interval has elapsed. Remove agent from list.
-                    self.state.remove(agent_id);
-                    false
-                } else {
-                    // Cool down interval has not elapsed, still cooling down.
-                    true
-                }
+            Some((instant, exponent)) => {
+                instant.elapsed().as_millis()
+                    < (self.back_off_interval * 2_u64.pow(*exponent)) as u128
             }
             None => false,
         }
     }
+
+    pub fn is_agent_at_max_back_off(&self, agent_id: &AgentId) -> bool {
+        self.state
+            .get(agent_id)
+            .map(|v| v.1 == self.max_back_off_exponent)
```

Comment:
Slightly confusing here too, because we'll actually only wait for delay * 2^(exponent - 1) as the last back off. The agent will report being at max back off before they have necessarily waited for the whole period. I think the logic actually works here but I'd be more confident if this was a check that the final window had expired. 

Though, see the comment above about a library that does this?

Reply:
Fixed name and also check of whether the back off interval has elapsed.